api/src/main/java/org/apache/iceberg/types/Types.java

```diff
@@ -543,6 +563,122 @@ public int hashCode() {
     }
   }
 
+  public static class GeometryType extends PrimitiveType {
+
+    private final String crs;
+
+    private GeometryType(String crs) {
+      Preconditions.checkNotNull(crs, "CRS cannot be null");
+      this.crs = crs;
+    }
+
+    public static GeometryType get() {
+      return new GeometryType("");
+    }
+
+    public static GeometryType of(String crs) {
+      return new GeometryType(crs == null ? "" : crs);
+    }
+
+    @Override
+    public TypeID typeId() {
+      return TypeID.GEOMETRY;
+    }
+
+    public String crs() {
+      return crs;
+    }
+
+    @Override
+    public boolean equals(Object o) {
+      if (this == o) {
+        return true;
+      } else if (!(o instanceof GeometryType)) {
+        return false;
+      }
+
+      GeometryType that = (GeometryType) o;
+      return crs.equals(that.crs);
+    }
+
+    @Override
+    public int hashCode() {
+      return Objects.hash(GeometryType.class, crs);
+    }
+
+    @Override
+    public String toString() {
+      return String.format("geometry(%s)", crs);
+    }
+  }
+
+  public static class GeographyType extends PrimitiveType {
+
+    private final String crs;
+    private final EdgeInterpolationAlgorithm algorithm;
+
+    private GeographyType(String crs, EdgeInterpolationAlgorithm algorithm) {
+      Preconditions.checkNotNull(crs, "CRS cannot be null");
+      this.crs = crs;
+      this.algorithm = algorithm;
+    }
+
+    public static GeographyType get() {
+      return new GeographyType("", null);
+    }
+
+    public static GeographyType of(String crs) {
+      return new GeographyType(crs, null);
+    }
+
+    public static GeographyType of(String crs, EdgeInterpolationAlgorithm algorithm) {
+      return new GeographyType(crs, algorithm);
+    }
+
+    public static GeographyType of(String crs, String algorithmName) {
+      EdgeInterpolationAlgorithm algorithm =
+          ((algorithmName == null || algorithmName.isEmpty())
+              ? null
+              : EdgeInterpolationAlgorithm.fromName(algorithmName));
+      return new GeographyType(crs == null ? "" : crs, algorithm);
+    }
+
+    @Override
+    public TypeID typeId() {
+      return TypeID.GEOGRAPHY;
+    }
+
+    public String crs() {
+      return crs;
+    }
+
+    public EdgeInterpolationAlgorithm algorithm() {
+      return algorithm;
+    }
+
+    @Override
+    public boolean equals(Object o) {
+      if (this == o) {
+        return true;
+      } else if (!(o instanceof GeographyType)) {
+        return false;
+      }
+
+      GeographyType that = (GeographyType) o;
+      return Objects.equals(crs, that.crs) && Objects.equals(algorithm, that.algorithm);
+    }
+
+    @Override
+    public int hashCode() {
+      return Objects.hash(GeographyType.class, crs, algorithm);
+    }
+
+    @Override
+    public String toString() {
+      return String.format("geography(%s, %s)", crs, algorithm != null ? algorithm.value() : "");
```

Comment:
Looks like this will produce `"geography(, )"` for the default type. When the type is the default, it should produce just `geography`. When algorithm is not set, it should produce `geography(crs)` and when algorithm is set it should produce `geography(crs, algorithm)`. When the CRS is defaulted and algorithm is not, it should use whatever standard name we decide on for CRS84.

Reply:
The problem is that there's no standard name for CRS84. `OGC:CRS84` mentioned in the spec is not a valid value for CRS. I suggest that we change the spec to allow `OGC:CRS84` to represent the default CRS. @jiayuasu @szehon-ho

Reply:
I think the spec says 'ogc:crs84' is default, and 'srid' and 'projjson' is for extensions, so its ok as a default value for me.  Wdyt @jiayuasu ?

Reply:
I am using `OGC:CRS84` for CRS when only algorithm is specified. One consequence is that when the serialized type is parsed back to `GeographyType` object, the `crs` field will be `OGC:CRS84` instead of `null`. I'm not sure if we could leave such behavior as is.